### ISSUE TITLE
[FEATURE] Use own stats instead of ethosdistro.com

### ISSUE THREAD

**Issue by jmverges** [foraern/ethos-panel#10]
Hi there,

I like the panel a lot, however, I would suggest to use own stats instead of the ones delivered by ethosdistro.com. I think could be easy to achieve this hooking into the opt/ethos/sbin/ethos-stats-daemon a post to some endpoint into this panel and some "token" generated in this panel to allow the post.

Cheers

**Comment by mvnuland** [comment]
Very good idea. I'd also like to see something like this.

**Comment by ricardo-nolan** [comment]
Only problem and reason I haven't done this yet is that a lot of users aren't sufficiently confortable with linux to be able to change ethos to do this. 
An alternative is to create a separate script to run on a cronjob and do the push, but again, people would need to install it on each rig and configure the cronjob..

**Comment by Overseer486** [comment]
well, that's do-able though
I'd be interested in getting stats directly from ethos as well!

**Comment by jmverges** [comment]
@foraern go for the second option then, a separate script will be easier for non linux users as that could be a one liner which download and install the cron for you, like with ohmyzsh or composer.

**Comment by Trbb** [comment]
i m a noob linux user  :(

**Comment by dogelyfe** [comment]
Just discovered this, added panel-- does this work as intended still, or do we need to go with the direct-from-rig route?

I've been digging and I found this code that may be similar to what you'd look to achieve?

http://adrian.ihacks.ro/ethos-custom-panel/

Not too familiar with Linux, but the tutorial prescribed seems simple enough.

**Comment by ricardo-nolan** [comment]
@arsoniztonfyre Ethos-panel itself works as intended, however, I think people who do not have a paid copy of ethosdistro may not get stats on it since they've recently made changes.

I'm investigating options for the direct-from-rig route.

The link you provided is what I'm trying to avoid, since it requires much more custom work from users rather than simply using a cronjobbed script.



There's also the ethical question I'm pondering, in that by using the direct-from-rig route, I may be facilitating people who "pirate" ethosdistro.

**Comment by ricardo-nolan** [comment]
So the question I have to ponder is:

Are people paying ethosdistro for the distribution itself, or just for the use of the dashboard on ethosdistro.com?

If we assume that people are paying for the distribution, then I'd be facilitating piracy.

If we assume that people are only paying for the use of the online dashboard, then I'm simply giving them another free option.

**Comment by dogelyfe** [comment]
Kotarius himself (ethOS founder) said users are paying for continued development and support rather than the panel, and "Our team does not mind that you create a panel. Personally, I encourage it. ":
https://forum.ethereum.org/discussion/11871/do-you-guys-want-an-ethos-custom-web-stats-panel

And I agree they have spent a lot of time and support for their paying clients. On the other hand ethOS could be in violation of licensing by distributing for profit:
https://www.reddit.com/r/EtherMining/comments/7fa89s/you_should_report_ethos_for_violating_the_gnu_gpl/

I'm not a lawyer.

**Comment by ricardo-nolan** [comment]
Seems, fair, since Kotarius himself is fine with it, I have no arguments.

I'll see if I can get something whipped up.

**Comment by ricardo-nolan** [comment]
Looks like the easiest way to do this will be to simply push /var/run/ethos/stats.json to a receiving process on ethos-panel.

**Comment by danielbowne** [comment]
Nice foraern.  Thanks for the work.

**Comment by ricardo-nolan** [comment]
Quick update:

After delving into this a bit, I think this is the best way to move forward:
1. Push /var/run/ethos/stats.json via curl on a crontabbed job to ethos-panel.
2. On ethos-panel, a new section will be added to the profile page to allow users to input their rig's public IP for whitelisting.
3. When ethos-panel receives a push from the rig, it will check if the IP is whitelisted, and identify user ID based on the IP, then insert the data into the db.
4. If the IP is not whitelisted, it ignores the push.
5. Users without an IP in the whitelist will pull stats from ethosdistro.com as before.

Thoughts?

**Comment by Overseer486** [comment]
Hmmmm
what about us with dynamic ip ?
could it work with dynamic dns?

**Comment by ricardo-nolan** [comment]
I'd have to test it, but generally speaking, nowadays a dynamic IP is fairly static

I am open to other methods of authentication though.

**Comment by Overseer486** [comment]
I know of some providers that regularly change the ip usually once a few days / week.

A method would be to have a string somewhere in your profile and have the rigs send a string with every push.
I'm not sure how this would be implemented though, the ip solution if by far simpler.

Or have the rigs login and then push the data?
Ah, i dunno

**Comment by ricardo-nolan** [comment]
well, we're not exactly dealing with super sensitive data, all the rig will be sending is a file with stats.

we could potentially use a url like: https://ethos-panel.com/post.php?id=yourdashboardid

where dashboard id is the six character string you currently use on ethosdistro.com (ie. xxxxxx.ethosdistro.com)


It's possible someone else could send stats for your dashboard, but not sure there's a point.

I could also implement basic authentication on the curl request, using your normal ethos-panel login,  but considering a lot of people don't know how to work with linux, it would add to the complexity for users

**Comment by mvnuland** [comment]
I know some miners (including me) that are working with a 4G connection on their farm, so we don't have a static IP.

When you know someone's panel ID you are also able to see the public data on ethosdistro, so that will be no problem for me. I also think the basic Linux users have to make some effort, otherwise this project doesn't fit them.

**Comment by danielbowne** [comment]
> "I'd have to test it, but generally speaking, nowadays a dynamic IP is fairly static"

Try to do in a way that allows IP whitelisting and fqdn for people using ddns.

**Comment by jmverges** [comment]
I would suggest as easier approach to use a token generated in the panel. Then, everything that post with that token will be accepted. Easier than ips, fqdn, etc

**Comment by ricardo-nolan** [comment]
Token idea might be the best way.

Possibly I'll simply add radio button to select between ethosdistro.com or direct rig stats, and add in the following for people to run on their rig:

`(crontab -l 2>/dev/null; echo "*/10 * * * * curl -X POST -H "Content-Type: multipart/form-data" -F 
 "data=@/var/run/ethos/stats.json;token=XXXXX" https://ethos-panel.com/post.php") | crontab -`

**Comment by ricardo-nolan** [comment]
I think that's now working. If you guys want to try it out and let me know if you find any bugs, I'd appreciate it.

Config for it is in the profile page.

**Comment by ricardo-nolan** [comment]
Found a bug on the post from rig. Will sort it asap

**Comment by ricardo-nolan** [comment]
bug resolved. 

had ; in the curl request which need to be removed. So updated profile page from:

`(crontab -l 2>/dev/null; echo '*/10 * * * * curl -X POST -H "Content-Type: multipart/form-data" -F "token=xxxxxxxxxxxx;" -F "data=@/var/run/ethos/stats.json;" https://ethos-panel.com/post.php') | crontab -`

to

`(crontab -l 2>/dev/null; echo '*/10 * * * * curl -X POST -H "Content-Type: multipart/form-data" -F "token=xxxxxxxxxxxx" -F "data=@/var/run/ethos/stats.json" https://ethos-panel.com/post.php') | crontab -`

Would appreciate if others can test the new features.

**Comment by Trbb** [comment]
sorry ... i m a noob linux user and english not good :(
on Ethos Panel i select Direct from Rig
and page tell me  run the following command on ecah rig
so i try this with 1 rig is mean use ssh and login rig and copy
 (crontab -l 2>/dev/null; echo '*/10 * * * * curl -X POST -H "Content-Type: multipart/form-data" -F "token=0581c5b39c0f79570011" -F "data=@/var/run/ethos/stats.json" https://ethos-panel.com/post.php') | crontab -

this command ? no need to change other ... i copy all and run but ethos panel still no rig

thanks for help and if u have zec wallect i can donations

**Comment by ricardo-nolan** [comment]
correct, just log into your rig via ssh and run that command, that's all.

Once that is done, it will update ethos-panel.com every 10 minutes.

**Comment by Trbb** [comment]
ok so i do right step
but on my rig after i run command no any message , i mean if correct must have some message right?
and my ethos panel still no rig

**Comment by Overseer486** [comment]
Yup, same problem here. There is a sort delay until the command is executed (uploading file maybe?) but there is no output and no update on ethos-panel.com

**Comment by ricardo-nolan** [comment]
correct, running the command above just adds it to your cron jobs, and will only be run 10 minutes later.

**Comment by Overseer486** [comment]
ah, sorry, i meant that i ran the command manually, by executing

curl -X POST -H "Content-Type: multipart/form-data" -F "token=XXXXXX" -F "data=@/var/run/ethos/stats.json" https://ethos-panel.com/post.php

and then:
There is a sort delay until the command is executed (uploading file maybe?) but there is no output and no update on ethos-panel.com

I also added the command to the crontab of 8 rigs about half an hour ago but still no update to the panel.

**Comment by ricardo-nolan** [comment]
@Overseer486 I see a single push from what looks like a curl for rig 4a8ad4:

23834744	208	2018-06-08 10:51:00	4a8ad4	169.56

I can tell it's a push because the standard request is always at the 10 minute mark, and this one ran at 51 minutes.

I don't see requests from other rigs. (I'll double check to ensure there's no overwrite problems).

**Comment by ricardo-nolan** [comment]
at 11:00 I see this:

23835240	208	2018-06-08 11:00:00	4a8ad4	169.57

So I assume you switched it back to ethosdistro.com ?

**Comment by ricardo-nolan** [comment]
could simply be that the command is running precisely at the 10 minute mark too

**Comment by Overseer486** [comment]
The one ran at 10:51:00 i did manually, anything else you see is from cron jobs.
But i still see nothing in my panel :/

Let me know if there's anything you'd like me to check!

**Comment by ricardo-nolan** [comment]
> But i still see nothing in my panel :/

Can you clarify that? Do you mean your stats completely stop at a certain time?

**Comment by Trbb** [comment]
did u see my rig on?
7a3cf4

**Comment by Overseer486** [comment]
@foraern No, i mean my Ethos-panel dashboard shows no rigs at all.

**Comment by Trbb** [comment]
@Overseer486  
ummm... my run command is not same to u 
i use 
(crontab -l 2>/dev/null; echo '*/10 * * * * curl -X POST -H "Content-Type: multipart/form-data" -F "token=XXXXXXXXXX" -F "data=@/var/run/ethos/stats.json" https://ethos-panel.com/post.php') | crontab -

but ur command is no have "(crontab -l 2>/dev/null; echo '*/10 * * * *"

so no need "(crontab -l 2>/dev/null; echo '*/10 * * * *"  right ? sorry i m a noob user

**Comment by jianyi85** [comment]
Hi foraern,

I also do the same thing, the panel never update, I am not sure the script is working or not.

Can you check my rigs hostname, see have push the data to your server or not...

05e76c
05e069
054b17
091d68
0917ee
5f7f5b

**Comment by ricardo-nolan** [comment]
@jianyi85 I see the following in the db:

23846265	229	2018-06-08 13:10:00	5f7f5b	177.37
23845788	229	2018-06-08 13:09:00	0917ee	3242.77
23845787	229	2018-06-08 13:09:00	05e069	2816.85
23845786	229	2018-06-08 13:08:00	05e76c	3241.05
23845785	229	2018-06-08 13:08:00	054b17	3227.75
23845252	229	2018-06-08 13:00:00	5f7f5b	177.43
23844925	229	2018-06-08 12:59:00	0917ee	3247.48
23844926	229	2018-06-08 12:59:00	05e069	2826.71
23844924	229	2018-06-08 12:58:00	05e76c	3237.13
23844923	229	2018-06-08 12:58:00	054b17	3240.91
23844395	229	2018-06-08 12:50:00	5f7f5b	177.40
23844069	229	2018-06-08 12:49:00	0917ee	3243.85
23844068	229	2018-06-08 12:49:00	05e069	2822.35
23844067	229	2018-06-08 12:48:00	05e76c	3219.62
23844066	229	2018-06-08 12:48:00	054b17	3232.19
23843536	229	2018-06-08 12:40:00	5f7f5b	177.40
23843212	229	2018-06-08 12:39:00	0917ee	3250.15
23843211	229	2018-06-08 12:39:00	05e069	2821.05
23843210	229	2018-06-08 12:38:00	05e76c	3251.35
23843209	229	2018-06-08 12:38:00	054b17	3240.77
23843208	229	2018-06-08 12:34:00	5f7f5b	177.43
23842675	229	2018-06-08 12:30:00	5f7f5b	177.40

I presume it's because you've set them up with a few minutes apart, but data is definitely coming in.

Can you logout of the panel, log back in and then check if it's still not updating please?

**Comment by ricardo-nolan** [comment]
@Overseer486 ah, ok, I think we're getting to the bottom of it now. Give me a few minutes.

**Comment by jianyi85** [comment]
@foraern

No, the data nvr show in the panel.

**Comment by Trbb** [comment]
@foraern  sir is have my rig? host name 7a3cf4 thx

**Comment by ricardo-nolan** [comment]
@jianyi85 I know what it is now, just need to figure out how to fix it.

**Comment by ricardo-nolan** [comment]
@Trbb I don't see your rig in the database.

**Comment by ricardo-nolan** [comment]
@Overseer486 @jianyi85 that should be working now, can you test it again when able?

**Comment by Overseer486** [comment]
@foraern I think it only shows the last rig that pinged the panel.

I was seeing 820235 as my only rig a while ago, and now i see that i still have one rig, the 4a8ad4

**Comment by Trbb** [comment]
so hard...
which step is fail
1.on Ethos Panel  web site select Direct from Rig
2.and copy after select display red command  ex :(crontab -l 2>/dev/null; echo '*/10 * * * * curl -X POST -H "Content-Type: multipart/form-data" -F "token=XXXXXXXXX" -F "data=@/var/run/ethos/stats.json" https://ethos-panel.com/post.php') | crontab -

3.all copy and past to rig and run
4.this is all my step
or need install something before select Direct from Rig?
plz help me ... :(

**Comment by Overseer486** [comment]
@Trbb you're good to go.
foraern is still working on it though be a little patient :)

**Comment by ricardo-nolan** [comment]
yep, fixing it now to check the correct number of rigs

please bear with me, still a number of fixes in progress, this was a bit of a rush (let's see if it works) job.

**Comment by jianyi85** [comment]
@foraern 
Yes, the panel can display the rigs information now. Nice work! :)

**Comment by Trbb** [comment]
@Overseer486  sorry my english is bad so i use google translation to talk
i have patient, and i just do not understand why my rig not show on his server ..  :(

**Comment by mvnuland** [comment]
The panel shows my rigs now.

I would like to see some more info in the future, like the card type and memory brand, but for now very nice work!

**Comment by ricardo-nolan** [comment]
ok, think all the bugs are sorted now. Please let me know if you find any.
I'll leave the issue open until next week, and if no other bugs are found, I'll close it then.

**Comment by ricardo-nolan** [comment]
@mvnuland we have all the information in /var/run/ethos/stats.json available to us. So just let me know the name of the variable you'd like to see, and I can add it in.

Keep in mind we have limited screen space.

**Comment by ricardo-nolan** [comment]
@Trbb I've checked the db, and your user has never had data sent to the server, either via ethosdistro.com or via the direct to rig push.

Can you copy paste the output of your ssh console when you run this:

curl -X POST -H "Content-Type: multipart/form-data" -F "token=0581c5b39c0f79570011" -F "data=@/var/run/ethos/stats.json" https://ethos-panel.com/post.php

**Comment by danielbowne** [comment]
Working now @foraern .  Great work!

**Comment by ricardo-nolan** [comment]
hmmm...something to consider:

because you'll be inputting the curl/cronjob manually on each rig, it is likely that the push will be desyncronized. So you could look at the graph and have rig1 with a hash rate at 10:09, and rig 2 with no hashrate at 10:09, and then rig1 with no hash rate at 10:12 and rig 2 with hash rate at 10:12

I think the best bet is to change the cron to the following:

(crontab -l 2>/dev/null; echo '0,10,20,30,40,50 * * * * curl -X POST -H "Content-Type: multipart/form-data" -F "token=XXXXXXXx" -F "data=@/var/run/ethos/stats.json" https://ethos-panel.com/post.php') | crontab -

This way, both rigs would publish hashrates at the same time.

Thoughts?

**Comment by mvnuland** [comment]
My suggestions for now are: Split the rig and worker table so we can use it in ascending order (for example I have rigs with the name 1060-01, 1060-02, 1060-03 and so on). It will be nice if I can see them in a chronological order.

Also I'd like to see clickable internal IP's to use Shell In A Box very quickly. The total hashrate of all rigs together will also be nice (maybe this is possible in the hashrate bar).

I have to think about the other info because of the limited screen page. Maybe put the GPU info on another page? But I think this isn't in a hurry right now. The most important information is already there.

And about your last post; I think it will be better to push the data synced from all rigs when that's possible with some changes. Keep up the good work!

**Comment by Trbb** [comment]
@foraern this is i paste and run picture https://imgur.com/HlGIlmU

**Comment by ricardo-nolan** [comment]
@Trbb can you run *just* the following command and post the screenshot?

curl -X POST -H "Content-Type: multipart/form-data" -F "token=0581c5b39c0f79570011" -F "data=@/var/run/ethos/stats.json" https://ethos-panel.com/post.php

**Comment by ricardo-nolan** [comment]
@mvnuland I've updated the cron string in profile page so it synchs.

your other requests I'll see if I can get to them this weekend (but more than likely it'll be next week).

**Comment by Trbb** [comment]
@foraern ok this is curl -X POST -H“Content-Type：multipart / form-data”-F“token = 0581c5b39c0f79570011”-F“data=@/var/run/ethos/stats.json” https://ethos-panel.com /post.php
paste and run picture
1.https://imgur.com/JLpEDSk
2.https://imgur.com/6vzU8z8
3.https://imgur.com/i92wgMJ

**Comment by Overseer486** [comment]
@foraern great work m8!
I agree with @mvnuland, i'd also like to see more stats on the panel, like bios/gpu/ram/volt/watt etc. etc. 
Or they could be on a separate panel page.
They really speed up rig manegement.

Once again, thank you for the great work!

**Comment by Trbb** [comment]
@foraern but u give me command can run with eth rig...i m no idea why can not run with my zec rig

eth rig=https://imgur.com/POxIniT

**Comment by ricardo-nolan** [comment]
@Trbb I have the following info:

```
23862915	216	2018-06-08 16:27:00	a70eff	177.42	
23864653	216	2018-06-08 16:43:00	7a3d37	177.00
```

Also, from your screenshots, on rig 7a3cf4 you ran the command incorrectly.

You ran the command as: 
`curl -X POST -H“Content-Type：multipart / form-data”-F“token = 0581c5b39c0f79570011”-F“data=@/var/run/ethos/stats.json” https://ethos-panel.com /post.php`

it should be:
`curl -X POST -H“Content-Type：multipart / form-data”-F“token = 0581c5b39c0f79570011”-F“data=@/var/run/ethos/stats.json” https://ethos-panel.com/post.php`

notice there's no space between ethos-panel.com and /post.php.

Once you have that working, we can look at sorting out your cronjob, since there's only the 2 entries in the DB, I suspect the cronjob isn't set up.

**Comment by ricardo-nolan** [comment]
For what it's worth, I noticed that I couldn't get my rigs to push at the same time, and noticed one of my rig's had the clock 40 seconds faster than the other.

the solution is to do the following on each rig:

```
sudo apt-get-ubuntu install ntp
sudo service ntp start
```

that should synchronize the clocks to the internet and they should start pushing at the same time (may take a few minutes for each rig to set the right time)

**Comment by ricardo-nolan** [comment]
Will close this issue tomorrow if nothing else is raised.

**Comment by Overseer486** [comment]
I just noticed some of my rigs went down due to internet problems a few hours ago.
The downed ones just vanished from the rig list.

Shouldn't they linger there for a couple of days before disappearing?

**Comment by ricardo-nolan** [comment]
Hadn't really considered that tbh, they're present in the graph, but in the rig list right now it only shows active rigs.
I'll have a think about it.

**Comment by jianyi85** [comment]
Having problem at this moment, the panel cannot save from profile.

**Comment by danielbowne** [comment]
`Having problem at this moment, the panel cannot save from profile.`
@jianyi85 having same issue.  I just manually saved settings in mysql.  Open users and set data origins to 1.

**Comment by ricardo-nolan** [comment]
Hi guys, problem with saving profile: on ethos-panel.com or self hosted?

If you’re using self hosted can you check your web server’s error log please, and post the relevant error here?

**Comment by jianyi85** [comment]
@foraern 

on ethos-panel.com, the profile was unable to save, totally not working.

**Comment by danielbowne** [comment]
Not working on self hosted for me.  I hard saved the settings in MYSQL manually.

**Comment by danielbowne** [comment]
Remote Config also not saving in self hosted.  No errors in Apache logs.

**Comment by ricardo-nolan** [comment]
Ah ok, I see the problem. Should have it sorted soon.

**Comment by ricardo-nolan** [comment]
Hi guys, saving profile should be working now, can one of you try it out?

**Comment by ricardo-nolan** [comment]
remote profile saving should now be working as well

**Comment by jianyi85** [comment]
@foraern 

Saving profile is working now, but saving the "Email Notifications" still not working.

**Comment by ricardo-nolan** [comment]
@jianyi85 that should be fixed now, sorry, had only had one coffee.

**Comment by ricardo-nolan** [comment]
Closing this issue, as the features are now implemented. Any further bugs found, please create a new issue.
